### PR TITLE
Add support for headers in action return

### LINKED ISSUE #2420: VirtualServerRoute: action.return headers
**Is your feature request related to a problem? Please describe.**
There are situations when using VirtualServerRoute's that you want to return a header when you use action.return, an example of this is shifting OPTIONS method requests into Nginx instead of the application (e.g. if you're pre-building ingresses via a helm chart) and you need to return the headers in an action.return - currently this is unsupported 

**Describe the solution you'd like**
Strangely, headers can be returned in the ErrorPage.return ( https://docs.nginx.com/nginx-ingress-controller/configuration/virtualserver-and-virtualserverroute-resources/#errorpagereturnheader ) - essentially this same mechanism can be used for action.return and then makes these consistent since they're almost identical. I also don't believe this an nginx.conf limitation.

An example and use case of what this might look like;

```yaml
       - conditions:
         - variable: $request_method
           value: "OPTIONS"
         action:
           return:
            code: 200
            body: "OK"
             headers:
             - name: Access-Control-Allow-Origin
               value: "*"
```

**Describe alternatives you've considered**
In theory, it looks like you'd be able to do this with the location snippet support but it seems like that goes against the conditional match patterns that VirtualServer brings to the table, e.g. you could return in the http snippet before the conditional match happens and there are some situations where you may want to do this (such as conditionals on other variables).

**Additional context**
I'd be happy to open a PR and start working on this if it's something that would be accept into the spec

**Update:** It looks like there's a bug in the location snippet whereby it doesn't get sent to the default routes so this doesn't work, I hadn't realised that default routes are sent through the ErrorPages struct with a default 418 code which then gets remapped, this looks to be causing issues with using location snippets here. The true workaround here (and is not great) is to create a service in the cluster that always returns "OK" and proxy pass that upstream until headers can be returned in action.return. So to be clear, snippets _do not_ work here.

Aha! Link: https://nginx.aha.io/features/IC-413

### PR DESCRIPTION
**Pull request by andrew-s**
### Proposed changes

This PR adds support for headers in the action return object. This closes/fixes #2420 

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

# Notes

I've added in unit tests, but it looks like there are other tests within the repo - I have also validated by compiling and running on a local cluster.

### CODE REVIEW

**Review by ADubhlaoich — COMMENTED** [review]
LGTM docs-wise but the majority of the PR is code, so I will not approve.

**Comment by andrew-s** [comment]
I've updated the PR to passing unit tests

**Comment by andrew-s** [comment]
@ADubhlaoich @haywoodsh

I think this needs someone from nginx to weigh in. With the PR as it is now, the generated/expected spec isn't correct - it would generate errorPages as;

```yaml
                              headers:
                                allOf:
                                - items:
                                    description: Header defines an HTTP Header.
                                    properties:
                                      name:
                                        type: string
                                      value:
                                        type: string
                                    type: object
                                - items:
                                    description: Header defines an HTTP Header.
                                    properties:
                                      name:
                                        type: string
                                      value:
                                        type: string
                                    type: object
                                type: array
```

This is because `ErrorPageReturn` uses `ActionReturn` which now has headers, so this applies headers twice. There's two solutions;

- ErrorPageReturn exclusively uses ActionReturn which now has headers, this would need all of the tests updating and anywhere that uses headers on ErrorPageReturn directly to then implement headers onto ActionReturn within the struct
- ErrorPageReturn no longer uses ActionReturn and implements `Code`, `Type` and `Body` directly - all of the tests and other logic which uses ActionReturn on ErrorPageReturn to then use these fields.

After some initial testing, these struct changes to produce the expected results.

I think these changes _could_ break any existing PRs but vice versa, if we decide one of these routes then it would make sense to try and merge quickly to prevent conflicts.

**Comment by haywoodsh** [comment]
Just trying to make it clearer, in the current API,`ActionReturn` has 3 fields, `Code`, `Type` and `Body`, and `ErrorPage` has 4 fields:`Code`, `Type`, `Body` and `Headers`, as it uses `ActionReturn` and `Headers`.  After this PR, both `Action.Return` and `ErrorPage.Return` will have the same items. 

I like the first solution more, because I don't think `Action.Return` and `ErrorPage.Return` will divert again in the future, like something should only exist for `Action` but not `ErrorPage`, or the other way around.

Also, I saw `ErrorPageRidirect` also has only one item `ActionRedirect`.
```
type ErrorPageRedirect struct {
	ActionRedirect `json:",inline"`
}
```
Therefore, I feel we should be consistent and remove `Headers` from `ErrorPageReturn` in this [line](https://github.com/nginxinc/kubernetes-ingress/blob/v3.5.0/pkg/apis/configuration/v1/types.go#L277), so that it becomes:

```
type ErrorPageReturn struct {
	ActionReturn `json:",inline"`
}
```

 As we are not changing the API for `ErrorPage.Return`, we might only need some minor refactorings to the unit tests.

**Comment by andrew-s** [comment]
That makes the most sense, I've made the change, updated tests and generated the new config.

**Comment by vepatel** [comment]
>  the unit tests can't check if the headers have actually been set

@haywoodsh @andrew-s I think we can add some assertions in following tests:

- https://github.com/andrew-s/kubernetes-ingress/blob/feature/action-return-headers/tests/suite/test_virtual_server_canned_responses.py ([data source](https://github.com/andrew-s/kubernetes-ingress/tree/feature/action-return-headers/tests/data/virtual-server-canned-responses))
- https://github.com/andrew-s/kubernetes-ingress/blob/feature/action-return-headers/tests/suite/test_v_s_route_canned_responses.py ([data source](https://github.com/andrew-s/kubernetes-ingress/tree/feature/action-return-headers/tests/data/virtual-server-route-canned-responses))

**Comment by andrew-s** [comment]
@haywoodsh @vepatel Added the suite tests, this should cover the three scenarios for both `virtualserver` and `virtualserverroute`;

- Invalid OpenAPI test
- Valid header and contents
- Update of header and contents

File: docs/content/configuration/virtualserver-and-virtualserverroute-resources.md [hunk 1]
```diff
@@ -592,10 +595,27 @@ return:
 |``code`` | The status code of the response. The allowed values are: ``2XX``, ``4XX`` or ``5XX``. The default is ``200``. | ``int`` | No |
 |``type`` | The MIME type of the response. The default is ``text/plain``. | ``string`` | No |
 |``body`` | The body of the response. Supports NGINX variables*. Variables must be enclosed in curly brackets. For example: ``Request is ${request_uri}\n``. | ``string`` | Yes |
+|``headers`` | The custom headers of the response. | [Action.Return.Header](#actionreturnheader) | No |
```
[the remark this answers is not in the document]
Comment (andrew-s) [reply docs/content/configuration/virtualserver-and-virtualserverroute-resources.md]: Yeah, this now follows the other examples, updated both

File: docs/content/configuration/virtualserver-and-virtualserverroute-resources.md [hunk 2]
```diff
@@ -592,10 +595,27 @@ return:
...
 
 \* -- Supported NGINX variables: `$request_uri`, `$request_method`, `$request_body`, `$scheme`, `$http_`, `$args`, `$arg_`, `$cookie_`, `$host`, `$request_time`, `$request_length`, `$nginx_version`, `$pid`, `$connection`, `$remote_addr`, `$remote_port`, `$time_iso8601`, `$time_local`, `$server_addr`, `$server_port`, `$server_name`, `$server_protocol`, `$connections_active`, `$connections_reading`, `$connections_writing` and `$connections_waiting`.
 
+### Action.Return.Header
+
+The header defines an HTTP Header for a canned response in an errorPage:
+
+```yaml
+name: x-debug-original-statuses
+value: ${upstream_status}
```
[the remark this answers is not in the document]
Comment (andrew-s) [reply docs/content/configuration/virtualserver-and-virtualserverroute-resources.md]: You're right, this example didn't really make any sense so I've updated it to follow the full example.

**Review by jjngx — APPROVED** [review]
🚀

**Review by haywoodsh — APPROVED** [review]

**Comment by haywoodsh** [comment]
Thank you for your contribution @andrew-s!

**Pull request merged**